Dfplayer won't compile after 2025.11.0

### The problem

Since 2025.11.0 none of my esphome devices that contain a dfplayer mini have been able to compile.  I believe it is related to a breaking change in 2025.11 but I don't believe I can correct this from my end.  I have tried "Clean All" but issue persists.  Anticipating the dfplayer component needs to be updated and/or associated documentation needs to be correct for new syntax.  

### Which version of ESPHome has the issue?

2025.11.various & 2025.12.0

### What type of installation are you using?

Home Assistant Add-on

### What platform are you using?

ESP8266

### Component causing the issue

dfplayer

### YAML Config

```yaml
esphome:
  name: doorchime
esp8266:
  board: nodemcuv2
 
wifi:
  ssid: !secret wifi_ssid
  password: !secret wifi_password
  reboot_timeout: 0s

# Enable logging
logger:
#  level: VERBOSE

# Example configuration entry
uart:
  tx_pin: D4
  rx_pin: D3
  baud_rate: 9600

# Declare DFPlayer mini module
dfplayer:
  on_finished_playback:
    then:
      logger.log: 'Somebody press play!'

# Enable Home Assistant API
api:
  reboot_timeout: 0s
  encryption:
    key: REMOVED
  services:
  - service: dfplayer_next
    then:
      - dfplayer.play_next:
  - service: dfplayer_previous
    then:
      - dfplayer.play_previous:
  - service: dfplayer_play
    variables:
      file: int
    then:
      - dfplayer.play: !lambda 'return file;'
  - service: dfplayer_play_loop
    variables:
      file: int
      loop_: bool
    then:
      - dfplayer.play:
          file: !lambda 'return file;'
          loop: !lambda 'return loop_;'
  - service: dfplayer_play_folder
    variables:
      folder: int
      file: int
    then:
      - dfplayer.play_folder:
          folder: !lambda 'return folder;'
          file: !lambda 'return file;'

  - service: dfplayer_play_loo_folder
    variables:
      folder: int
    then:
      - dfplayer.play_folder:
          folder: !lambda 'return folder;'
          loop: True

  - service: dfplayer_set_device
    variables:
      device: int
    then:
      - dfplayer.set_device:
          device: TF_CARD

  - service: dfplayer_set_volume
    variables:
      volume: int
    then:
      - dfplayer.set_volume: !lambda 'return volume;'
  - service: dfplayer_set_eq
    variables:
      preset: int
    then:
      - dfplayer.set_eq: !lambda 'return static_cast<dfplayer::EqPreset>(preset);'

  - service: dfplayer_sleep
    then:
      - dfplayer.sleep

  - service: dfplayer_reset
    then:
      - dfplayer.reset

  - service: dfplayer_start
    then:
      - dfplayer.start

  - service: dfplayer_pause
    then:
      - dfplayer.pause

  - service: dfplayer_stop
    then:
      - dfplayer.stop

  - service: dfplayer_random
    then:
      - dfplayer.random

ota:
  platform: esphome
  password: "REMOVED"
  
binary_sensor:
  - platform: gpio
    name: "Front Doorbell"
    pin:
      number: D2
      inverted: True
      mode: INPUT_PULLUP
    filters:
      - delayed_on: 30ms
      - delayed_off: 10ms
  - platform: gpio
    name: "Back Doorbell"
    pin:
      number: D1
      inverted: True
      mode: INPUT_PULLUP
    filters:
      - delayed_on: 30ms
      - delayed_off: 10ms
```

### Anything in the logs that might be useful for us?

```txt
INFO ESPHome 2025.12.0
INFO Reading configuration /config/esphome/doorchime.yaml...
WARNING The minimum WiFi authentication mode (wifi -> min_auth_mode) is not set. This controls the weakest encryption your device will accept when connecting to WiFi. Currently defaults to WPA (less secure), but will change to WPA2 (more secure) in 2026.6.0. WPA uses TKIP encryption which has known security vulnerabilities and should be avoided. WPA2 uses AES encryption which is significantly more secure. To silence this warning, explicitly set min_auth_mode under 'wifi:'. If your router supports WPA2 or WPA3, set 'min_auth_mode: WPA2'. If your router only supports WPA, set 'min_auth_mode: WPA'.
INFO Generating C++ source...
INFO Compiling app... Build path: /data/build/doorchime
********************************************************************************
If you like PlatformIO, please:
- star it on GitHub > https://github.com/platformio/platformio-core
- follow us on LinkedIn to stay up-to-date on the latest project news > https://www.linkedin.com/company/platformio/
- try PlatformIO IDE for embedded development > https://platformio.org/platformio-ide
********************************************************************************

Processing doorchime (board: nodemcuv2; framework: arduino; platform: platformio/espressif8266@4.2.1)
--------------------------------------------------------------------------------
Platform Manager: Installing platformio/espressif8266 @ 4.2.1
INFO Installing platformio/espressif8266 @ 4.2.1
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Platform Manager: espressif8266@4.2.1 has been installed!
INFO espressif8266@4.2.1 has been installed!
Tool Manager: Installing platformio/toolchain-xtensa @ ~2.100300.220621
INFO Installing platformio/toolchain-xtensa @ ~2.100300.220621
Downloading  [####################################]  100%          
Unpacking  [####################################]  100%          
Tool Manager: toolchain-xtensa@2.100300.220621 has been installed!
INFO toolchain-xtensa@2.100300.220621 has been installed!
Tool Manager: Installing platformio/framework-arduinoespressif8266 @ ~3.30102.0
INFO Installing platformio/framework-arduinoespressif8266 @ ~3.30102.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%          
Tool Manager: framework-arduinoespressif8266@3.30102.0 has been installed!
INFO framework-arduinoespressif8266@3.30102.0 has been installed!
Tool Manager: Installing platformio/tool-esptool @ <2
INFO Installing platformio/tool-esptool @ <2
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Tool Manager: tool-esptool@1.413.0 has been installed!
INFO tool-esptool@1.413.0 has been installed!
Tool Manager: Installing platformio/tool-esptoolpy @ ~1.30000.0
INFO Installing platformio/tool-esptoolpy @ ~1.30000.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Tool Manager: tool-esptoolpy@1.30000.201119 has been installed!
INFO tool-esptoolpy@1.30000.201119 has been installed!
Tool Manager: Installing platformio/tool-scons @ ~4.40801.0
INFO Installing platformio/tool-scons @ ~4.40801.0
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Tool Manager: tool-scons@4.40801.0 has been installed!
INFO tool-scons@4.40801.0 has been installed!
Library Manager: Installing esphome/noise-c @ 0.1.10
INFO Installing esphome/noise-c @ 0.1.10
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: noise-c@0.1.10 has been installed!
INFO noise-c@0.1.10 has been installed!
Library Manager: Resolving dependencies...
INFO Resolving dependencies...
Library Manager: Installing esphome/libsodium @ 1.10020.7
INFO Installing esphome/libsodium @ 1.10020.7
Downloading  [####################################]  100%
Unpacking  [####################################]  100%
Library Manager: libsodium@1.10020.7 has been installed!
INFO libsodium@1.10020.7 has been installed!
HARDWARE: ESP8266 80MHz, 80KB RAM, 4MB Flash
Dependency Graph
|-- ESP8266WiFi @ 1.0
|-- ESP8266mDNS @ 1.2
|-- noise-c @ 0.1.10
Compiling .pioenvs/doorchime/src/esphome/components/api/api_connection.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/api_frame_helper.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/api_frame_helper_noise.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/api_pb2.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/api_pb2_service.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/api_server.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/list_entities.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/proto.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/subscribe_state.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/api/user_services.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/binary_sensor/automation.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/binary_sensor/binary_sensor.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/binary_sensor/filter.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/dfplayer/dfplayer.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/esp8266/core.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/esp8266/gpio.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/esp8266/helpers.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/esp8266/preferences.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/esphome/ota/ota_esphome.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/gpio/binary_sensor/gpio_binary_sensor.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/logger/logger.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/logger/logger_esp8266.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/md5/md5.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/mdns/mdns_component.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/mdns/mdns_esp8266.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/network/util.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/ota/ota_backend.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/ota/ota_backend_arduino_esp8266.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/safe_mode/safe_mode.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/sha256/sha256.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/socket/lwip_raw_tcp_impl.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/socket/socket.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/uart/uart.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/uart/uart_component.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/uart/uart_component_esp8266.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/uart/uart_debugger.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/wifi/wifi_component.cpp.o
Compiling .pioenvs/doorchime/src/esphome/components/wifi/wifi_component_esp8266.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/application.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/color.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/component.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/component_iterator.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/controller_registry.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/entity_base.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/helpers.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/log.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/scheduler.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/time.cpp.o
Compiling .pioenvs/doorchime/src/esphome/core/util.cpp.o
Compiling .pioenvs/doorchime/src/main.cpp.o
Generating LD script .pioenvs/doorchime/ld/local.eagle.app.v6.common.ld
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/BearSSLHelpers.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/CertStoreBearSSL.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFi.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiAP.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiGeneric.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiGratuitous.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiMulti.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiSTA-WPS.cpp.o
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::PlayFileAction<int>':
src/main.cpp:561:70:   required from here
src/esphome/components/dfplayer/dfplayer.h:121:8: error: 'void esphome::dfplayer::PlayFileAction<Ts>::play(Ts ...) [with Ts = {int}]' marked 'override', but does not override
  121 |   void play(Ts... x) override {
      |        ^~~~
src/main.cpp: In function 'void setup()':
src/main.cpp:561:70: error: invalid new-expression of abstract class type 'esphome::dfplayer::PlayFileAction<int>'
  561 |   dfplayer_playfileaction_id = new dfplayer::PlayFileAction<int32_t>();
      |                                                                      ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:116:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::PlayFileAction<int>':
  116 | template<typename... Ts> class PlayFileAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::PlayFileAction<int, bool>':
/config/esphome/doorchime.yaml:48:78:   required from here
src/esphome/components/dfplayer/dfplayer.h:121:8: error: 'void esphome::dfplayer::PlayFileAction<Ts>::play(Ts ...) [with Ts = {int, bool}]' marked 'override', but does not override
  121 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:48:78: error: invalid new-expression of abstract class type 'esphome::dfplayer::PlayFileAction<int, bool>'
   48 |     then:
      |                                                                              ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:116:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::PlayFileAction<int, bool>':
  116 | template<typename... Ts> class PlayFileAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int, bool}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::PlayFolderAction<int, int>':
/config/esphome/doorchime.yaml:56:83:   required from here
src/esphome/components/dfplayer/dfplayer.h:138:8: error: 'void esphome::dfplayer::PlayFolderAction<Ts>::play(Ts ...) [with Ts = {int, int}]' marked 'override', but does not override
  138 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:56:83: error: invalid new-expression of abstract class type 'esphome::dfplayer::PlayFolderAction<int, int>'
   56 |     then:
      |                                                                                   ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:132:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::PlayFolderAction<int, int>':
  132 | template<typename... Ts> class PlayFolderAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int, int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::PlayFolderAction<int>':
/config/esphome/doorchime.yaml:64:76:   required from here
src/esphome/components/dfplayer/dfplayer.h:138:8: error: 'void esphome::dfplayer::PlayFolderAction<Ts>::play(Ts ...) [with Ts = {int}]' marked 'override', but does not override
  138 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:64:76: error: invalid new-expression of abstract class type 'esphome::dfplayer::PlayFolderAction<int>'
   64 |     then:
      |                                                                            ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:132:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::PlayFolderAction<int>':
  132 | template<typename... Ts> class PlayFolderAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::SetDeviceAction<int>':
/config/esphome/doorchime.yaml:72:72:   required from here
src/esphome/components/dfplayer/dfplayer.h:154:8: error: 'void esphome::dfplayer::SetDeviceAction<Ts>::play(Ts ...) [with Ts = {int}]' marked 'override', but does not override
  154 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:72:72: error: invalid new-expression of abstract class type 'esphome::dfplayer::SetDeviceAction<int>'
   72 |     then:
      |                                                                        ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:150:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::SetDeviceAction<int>':
  150 | template<typename... Ts> class SetDeviceAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::SetVolumeAction<int>':
/config/esphome/doorchime.yaml:78:72:   required from here
src/esphome/components/dfplayer/dfplayer.h:164:8: error: 'void esphome::dfplayer::SetVolumeAction<Ts>::play(Ts ...) [with Ts = {int}]' marked 'override', but does not override
  164 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:78:72: error: invalid new-expression of abstract class type 'esphome::dfplayer::SetVolumeAction<int>'
   78 |       volume: int
      |                                                                        ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:160:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::SetVolumeAction<int>':
  160 | template<typename... Ts> class SetVolumeAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h: In instantiation of 'class esphome::dfplayer::SetEqAction<int>':
/config/esphome/doorchime.yaml:85:64:   required from here
src/esphome/components/dfplayer/dfplayer.h:174:8: error: 'void esphome::dfplayer::SetEqAction<Ts>::play(Ts ...) [with Ts = {int}]' marked 'override', but does not override
  174 |   void play(Ts... x) override {
      |        ^~~~
/config/esphome/doorchime.yaml:85:64: error: invalid new-expression of abstract class type 'esphome::dfplayer::SetEqAction<int>'
   85 |       - dfplayer.set_eq: !lambda 'return static_cast<dfplayer::EqPreset>(preset);'
      |                                                                ^
In file included from src/esphome.h:21,
                 from src/main.cpp:3:
src/esphome/components/dfplayer/dfplayer.h:170:32: note:   because the following virtual functions are pure within 'esphome::dfplayer::SetEqAction<int>':
  170 | template<typename... Ts> class SetEqAction : public Action<Ts...>, public Parented<DFPlayer> {
      |                                ^~~~~~~~~~~
In file included from src/esphome/components/binary_sensor/filter.h:3,
                 from src/esphome/components/binary_sensor/binary_sensor.h:5,
                 from src/esphome/core/application.h:30,
                 from src/esphome/components/api/api_frame_helper.h:13,
                 from src/esphome/components/api/api_connection.h:5,
                 from src/esphome.h:3,
                 from src/main.cpp:3:
src/esphome/core/automation.h:260:16: note:     'void esphome::Action<Ts>::play(const Ts& ...) [with Ts = {int}]'
  260 |   virtual void play(const Ts &...x) = 0;
      |                ^~~~
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiSTA.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/ESP8266WiFiScan.cpp.o
Compiling .pioenvs/doorchime/lib67b/ESP8266WiFi/WiFiClient.cpp.o
*** [.pioenvs/doorchime/src/main.cpp.o] Error 1
========================= [FAILED] Took 54.09 seconds =========================
```

### Additional information

Appreciate if someone could look into this.  Let me know if anything else is needed from my end.

--- a/esphome/loader.py
+++ b/esphome/loader.py
@@ -187,7 +187,14 @@ def install_meta_finder(
 
 
 def install_custom_components_meta_finder():
+    # Remove before 2026.6.0
     custom_components_dir = (Path(CORE.config_dir) / "custom_components").resolve()
+    if custom_components_dir.is_dir() and any(custom_components_dir.iterdir()):
+        _LOGGER.warning(
+            "The 'custom_components' folder is deprecated and will be removed in 2026.6.0. "
+            "Please use 'external_components' instead. "
+            "See https://esphome.io/components/external_components.html for more information."
+        )
     install_meta_finder(custom_components_dir)
 
 

